### ISSUE TITLE
Test grids (CNN) pipeline with old deeprank paper's data - crystallography experiment

### ISSUE THREAD

**Issue by gcroci2** [DeepRank/deeprank-core#376]
We should test the entire pipeline with the old deeprank paper's data, in particular reproducing the [crystallographic PPIs's experiment](https://www.nature.com/articles/s41467-021-27396-0#:~:text=for%20subsequent%20analysis.-,Application%201%3A%20Detecting%20crystal%20artifacts,-X%2DRay%20crystallography
). 

Tentative tasks:
- Determine and list roughly old hyperparameters used during training, such as data augmentation, network, grid settings, etc
- Train the CNN used in the paper and obtain results at least comparable with the ones published in the paper
- Open debugging issues

**Comment by ntxxt** [comment]
Hello, 
I am trying to re-train with deeprankcore with interface data and I have a question. 
It seems like when taking a list of hdf5 files as input, GridDataset obj can only read the last hdf5 path in the list (thus len(GridDataset) is always 31). Does it mean all the training files have to be combined in a big hdf5 file?

**Comment by ntxxt** [comment]
yes, I am sure.

**Comment by gcroci2** [comment]
Can you send us in Slack some of the hdf5 files you're using and the script as well? @DaniBodor and I will look into this

**Comment by ntxxt** [comment]
Thank you. I will send you the script and three hdf5 files on slack.

**Comment by ntxxt** [comment]
After reviewing the DeeprankDataset class, it appears that the _create_index_entries(self) function is not appending the index_entries list properly across multiple HDF5 files. I have fixed it in my local installation. But you might want to check as well. 
https://github.com/DeepRank/deeprank-core/blob/7a824711f849f6f06711d9d0974a8fa8bbb2783d/deeprankcore/dataset.py#L128

**Comment by gcroci2** [comment]
Could you try to rerun the code in the branch in https://github.com/DeepRank/deeprank-core/pull/397 and let me know? It should work now @ntxxt

**Comment by gcroci2** [comment]
We discussed at today's weekly meeting that @ntxxt will report the comparison of the experiment before augmentation vs after augmentation, for training, validation, and testing sets. Also the timings will be reported.

**Comment by ntxxt** [comment]
**More detail on the data:** @gcroci2 
Training set: 4592 in total; (80% of MANY)
validation set: 1146 in total ; (20% of MANY)
Test set: 161 in total  (full DC set)

**Train and valid loss:**
Trained for 24 epoch, 17 hours on one A100 GPU, 40 CPU. Model saved at epoch 3.
![image](https://user-images.githubusercontent.com/102460988/227967265-7a123c66-84c2-4cf3-8a3e-d021bc054aff.png)

**Train/valid/test confusion matrix and accuracy:** 

**Train**
| measure | argumented | Not argumented |
| --- | --- | ---| 
| tp |67457 |2209|
| tn | 69758 | 2282|
| fp |2472 | 48| 
| fn | 2617| 51|
| Accuracy | 0.964|0.978|

**Valid**
| measure | argumented | Not argumented |
| --- | --- | ---| 
| tp | 16091| 518|
| tn | 16896 | 543|
| fp | 1084| 37| 
| fn |1455 | 48| 
| Accuracy |0.929 |0.926|

**Test**
| measure | argumented | Not argumented |
| --- | --- | ---| 
| tp | 1907| 61|
| tn | 1958 | 61|
| fp | 553| 20|
| fn | 573| 19|
| Accuracy |0.774 | 0.757|

**Comment by gcroci2** [comment]
Thanks :) @ntxxt Do the 17 hours refer to the augmented or non-augmented data experiment?
To me, it seems that it's overfitting from almost the beginning. Is the loss curve similar to one of the old experiment? @LilySnow @sonjageorgievska

**Comment by ntxxt** [comment]
It refers to the augmented one. @gcroci2

**Comment by ntxxt** [comment]
After running the experiment, the result is as follows,
Sadly the test accuracy is still low, but there is a bit more increase in accuracy after augmentation. (0.702 -> 0.760 (for the augmented test set); 0.702 -> 0.727 (for the not augmented test set). 

| Measure | Train | Valid | Test | Augmented_Train| Augmented_Valid | Augmented_Test|Augmented_Test(exclude augmentation)
| --- | --- | --- | --- | --- | --- | --- | ---|
| tp| 2223|556|61|67973|16987|1787|55|
| tn| 2287 | 570|52|71297|17841|2008|62|
| fp| 43|10|29|993|139|503|19|
| fn|39|10|19|2101|559|593|25|
|Accuracy| 0.982|  0.983|0.702|0.978|0.980|0.760|0.727|

**Comment by gcroci2** [comment]
Original dataset
trainset 4592
valid set 1147
testset 161

Summary of experiments and comparisons

**Network**: two 3D CNN/Max pooling blocks followed by two linear fully connected layers. This results in 117,362 learnable parameters. _Verify the same was used for deeprankcore_
augmentation_count: 30
**Grid settings**: GridSettings([30, 30, 30], [30.0, 30.0, 30.0])
**Features**: 20 pssm in deeprankcore (compared to 40 in deeprank, pssm separated by chainA and B)
**Batch size**: ?
**Learning rate**: 0.001 _Verify the same was used for deeprankcore_

NA = Not Augmented
A = Augmented

## Train
| Measure | deeprankcore_NA | deeprank_NA | deeprankcore_A | deeprank_A
| --- | --- | --- | --- | --- |
| Data | 4592 | 4592 | 142364 | 142352 |
| tp | 2223 | ? | 67973 | ? |
| tn | 2287 | ? | 71297 | ? |
| fp | 43 | ? | 993 | ? |
| fn | 39 | ? | 2101 | ? |
| Accuracy | 0.982 | ? | 0.978 | ? |
| Settings/Resources | ? | ? | 24 epochs, A100 GPU, 40 CPU | ?
| Best epoch | ? | ? | 3 | 2 | ?
| Training time | ? | ? | 17h | ? |

## Test
| Measure | deeprankcore_NA | deeprank_NA | deeprankcore_A | deeprank_A
| --- | --- | --- | --- | --- |
| Data | 161 | 161 | 161 | 161 |
| tp | | | | |
| tn | | | | |
| fp | | | | |
| fn | | | | |
| Accuracy | | | | |

**Comment by ntxxt** [comment]
Added one embedding layer to the original model, resulting in 118202 parameters, trained on the augmented training set for 20 epochs,  to be on the same page as deeprank, using the model saved at epoch 2 

|  measure |   training_set  |  test_set         |
| --- | --- | --- |
|   tp:  |        54800         |        79    |
|   tn:     |     54700         |        69   |
|   fp:   |       16444             |    12   |
   fn:     |     16376     |            1      |
Accuray:   |    0.770   |         0.918

Loss:
![image](https://user-images.githubusercontent.com/102460988/231508292-8940fb80-2099-4071-8fce-821ca73814d6.png)

**Comment by DaniBodor** [comment]
> Added one embedding layer to the original model, resulting in 118202 parameters, trained on the augmented training set for 20 epochs, to be on the same page as deeprank, using the model saved at epoch 2
> measure 	training_set 	test_set
> tp: 	54800 	79
> tn: 	54700 	69
> fp: 	16444 	12
> fn: 	16376 	1
> Accuray: 	0.770 	0.918
> 
> Loss: <img alt="image" width="400" src="https://user-images.githubusercontent.com/102460988/231508292-8940fb80-2099-4071-8fce-821ca73814d6.png">


So do I see correctly that bassically that did the trick and it's now outperforming deeprank? 
Would we expect it to outperform though? Wouldn't we expect very similar performance?

**Comment by DaniBodor** [comment]
Can this issue be closed?
@LilySnow , @ntxxt , @gcroci2

**Comment by gcroci2** [comment]
> Can this issue be closed? @LilySnow , @ntxxt , @gcroci2

I guess so. Also, I tested deeprankcore with grids and CNNs using the pMHCI 100k dataset, and no bugs were found so far (see the issue [#152](https://github.com/DeepRank/3D-Vac/issues/152) in 3D-Vac repo). So I'd say that the CNN pipeline works as well.
